lib/websocket/connection.js

```diff
@@ -0,0 +1,409 @@
+
+      return
+    } else if (frame.opcode === opcodes.CLOSE) {
+      // Upon either sending or receiving a Close control frame, it is said
+      // that _The WebSocket Closing Handshake is Started_ and that the
+      // WebSocket connection is in the CLOSING state.
+      ws[kReadyState] = states.CLOSING
+      ws[kClosingFrame] = frame
+      frame = undefined
+      return
+    }
+
+    // rsv bits are reserved for future use; if any aren't 0,
+    // we currently can't handle them.
+    if (frame.rsv1 || frame.rsv2 || frame.rsv3) {
+      failWebsocketConnection(ws[kController])
+      return
+    }
+
+    assert(frame.opcode === opcodes.TEXT || frame.opcode === opcodes.BINARY)
+
+    // If the frame comprises an unfragmented
+    // message (Section 5.4), it is said that _A WebSocket Message Has Been
+    // Received_
+
+    if (frame.terminated) {
+      // 1. If ready state is not OPEN (1), then return.
+      if (ws[kReadyState] !== states.OPEN) {
+        return
+      }
+
+      // 2. Let dataForEvent be determined by switching on type and binary type:
+      let dataForEvent
+
+      if (frame.opcode === opcodes.TEXT) {
+        // - type indicates that the data is Text
+        //      a new DOMString containing data
+        dataForEvent = new TextDecoder().decode(frame.data)
+      } else if (frame.opcode === opcodes.BINARY) {
+        if (ws[kBinaryType] === 'blob') {
+          // - type indicates that the data is Binary and binary type is "blob"
+          //      a new Blob object, created in the relevant Realm of the
+          //      WebSocket object, that represents data as its raw data
+          dataForEvent = new Blob([frame.data])
+        } else {
+          // - type indicates that the data is Binary and binary type is
+          //   "arraybuffer"
+          //      a new ArrayBuffer object, created in the relevant Realm of the
+          //      WebSocket object, whose contents are data
+          dataForEvent = new Uint8Array(frame.data).buffer
+        }
+      }
+
+      // 3. Fire an event named message at the WebSocket object, using
+      //    MessageEvent, with the origin attribute initialized to the
+      //    serialization of the WebSocket object’s url's origin, and the data
+      //    attribute initialized to dataForEvent.
+      fireEvent('message', ws, MessageEvent, {
+        origin: ws[kWebSocketURL].origin,
+        data: dataForEvent
+      })
+
+      frame = undefined
+    }
+  })
+}
+
+/**
+ * @see https://websockets.spec.whatwg.org/#feedback-from-the-protocol
+ * @see https://datatracker.ietf.org/doc/html/rfc6455#section-7.1.4
+ * @param {import('./websocket').WebSocket} ws
+ */
+function socketClosed (ws) {
+  const { [kResponse]: response } = ws
+
+  response.socket.on('close', () => {
+    const wasClean = ws[kReadyState] === states.CLOSING || isEstablished(ws)
+
+    /** @type {ReturnType<WebsocketFrame['parseCloseBody']} */
+    const { code, reason } = ws[kClosingFrame]
```

Comment:
`code` must be a valid status code. `reason` must be valid UTF-8.

Reply:
Is there an easy way to validate utf-8 without taking ws' implementation? Code is easy enough.

Reply:
There is https://github.com/websockets/utf-8-validate, https://github.com/micnic/uv, https://github.com/hcodes/isutf8.

Reply:
> Or hopefully we could eventually have an implementation inside of node, similar to toUSVString, which would benefit anyone using websockets.

A native implementation in core could use https://github.com/simdutf/simdutf. This is used by Bun and is the state-of-art in terms of performance, see https://github.com/websockets/utf-8-validate/pull/101. I did not merge that PR because the library has a lot of unneeded features. It would be great to add compile time flags to create a binary only with the wanted features.

Reply:
There's a possibility of simdutf getting added as a dependency to node, I'm talking to anonrig and he already has a fork with it added. He just didn't have a reason to add it in node.

Reply:
@lpinca @KhafraDev I opened the PR. Appreciate, if you can share your reasoning the pull request so that TSC members can have a better understanding of the need. https://github.com/nodejs/node/pull/45803

Reply:
I've just remembered that there is `TextDecoder` now in Node.js. It is possible to use it to "validate" UTF-8.

```js
const decoder = new TextDecoder('utf-8', { fatal: true });

try {
  decoder.decode(Buffer.from([0xf8]));
} catch {
  console.log('Invalid UTF-8');
}
```

It can also be used in a streaming fashion which is what the spec recommends.

```js
const decoder = new TextDecoder('utf-8', { fatal: true });

const chunks = [
  0xf0, 0x90, 0x8c, 0x88, // 𐍈
  0xe0, 0x80, 0xa0, // Overlong
  0xe2, 0x82, 0xac  // €
];

let string = ''

for (let i = 0; i < chunks.length; i++) {
  const chunk = chunks[i];
  const isLastChunk = i === chunks.length - 1;

  try {
    string += decoder.decode(
      Buffer.from([chunk]),
      { stream: !isLastChunk }
    );

    if (isLastChunk) console.log(string);
  } catch {
    console.log('Invalid UTF-8 detected at index %d', i + 1);
    break;
  }
}
```

The only problem is that the `fatal` option is not supported when ICU is disabled.

Reply:
TIL the fatal option exists.